File: Code/Discounting Automation.R
library(tidyverse)
library(readxl)
library(rio)
library(dplyr)
library(tidyr)
library(ChainLadder)
library(lubridate)
library(readr)
library(zoo)
library(ggplot2)

data_OS_2021 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                   sheet = "Raw Data_OS_21")
data_OS_2022 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                           sheet = "Raw_Data_OS_22")
data_OS_2023 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                           sheet = "Raw_Data_OS_23")
data_Paid_2022 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                             sheet = "Raw_Data_Paid_22")
data_nse_yield_curve_2021 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                                        sheet = "NSE_Yield_Curve_2021")
data_nse_yield_curve_2022 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                                        sheet = "NSE_Yield_Curve_2022")  
data_nse_yield_curve_2023 <- read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                                        sheet = "NSE_Yield_Curve_2023")

#Claim Indicator Count, sum by OS Amount and Percentage 2021
claim_indicator_count_2021 <- data_OS_2021 %>%
  group_by(`Claim Indicator`) %>%
  summarise(count = n(), total_OS_Amount = sum(`OS Amount`)) %>%
  mutate(percentage = total_OS_Amount / sum(total_OS_Amount) * 100)

#Claim Indicator Count 2022
claim_indicator_count_2022 <- data_OS_2022 %>%
  group_by(`Claim Indicator`) %>%
  summarise(count = n(), total_OS_Amount = sum(`OS Amount`)) %>%
  mutate(percentage = count / sum(count) * 100)

#OS Amount summary by Class 2022
claim_indicator_sum_2022 <- data_OS_2022 %>%
  filter(`Claim Indicator`=='E1')%>%
  group_by(`IRA Class`) %>%
  summarize(total_OS_Amount = sum(`OS Amount`))

#Paid Claims 2022
Paid_claims_indicator_count_sum_2022 <- data_Paid_2022 %>%
  group_by(`Claim Indicator`) %>%
  summarise(count = n(), total_Gross_Paid = sum(`Gross Paid`))

#By Claim Indicator
Paid_Final_claims_indicator_count_sum_2022 <- data_Paid_2022 %>%
  group_by(`Final Claim Indicator`) %>%
  summarise(count = n(), total_Gross_Paid = sum(`Gross Paid`)) 

#Paid Claims 2022
#By Loss Year
Paid_claims_Loss_year_count_sum_2022 <- data_Paid_2022 %>%
  filter(`Revised Claim Indicator`=='B2')%>%
  group_by(`Loss Year`) %>%
  summarise(count = n(), total_Gross_Paid = sum(`Gross Paid`))

#Experience Adjustment
#2022 - OS claims by Loss Year
OS_claims_by_Loss_Year_2022 <- data_OS_2022 %>%
  group_by(`Loss Year`) %>%
  summarise(total_OS_Amount_2022 = sum(`OS Amount`))

#2023 - OS claims by Loss Year
OS_claims_by_Loss_Year_2023 <- data_OS_2023 %>%
  group_by(`Loss Year`) %>%
  summarise(total_OS_Amount_2023 = sum(`OS Amount`))













#OS Claims by IRA CLASS
OS_claims_by_IRA_Class <- data_OS_2023 %>%
  group_by(`IRA Class`) %>%
  summarise(total_OS_Amount = sum(`OS Amount`))

#OS Claims by IRA CLASS for 2023
OS_claims_by_IRA_Class_2023 <- data_OS_2023 %>%
  filter(`Loss Year` == 2023) %>%
  group_by(`IRA Class`) %>%
  summarise(total_OS_Amount_2023 = sum(`OS Amount`))

File: Code/Discounting_Automation.py
import pandas as pd
import numpy as np

# Load the data from Excel sheets into pandas DataFrames
data_OS_2021 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                   sheet_name = "Raw Data_OS_21")
data_OS_2022 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                           sheet_name = "Raw_Data_OS_22")
data_OS_2023 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                           sheet_name = "Raw_Data_OS_23")
data_Paid_2022 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                             sheet_name = "Raw_Data_Paid_22")

# Load the NSE Yield Curve data from Excel sheets into pandas DataFrames
data_nse_yield_curve_2021 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                                        sheet_name = "NSE_Yield_Curve_2021")
data_nse_yield_curve_2022 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                                        sheet_name = "NSE_Yield_Curve_2022")  
data_nse_yield_curve_2023 = pd.read_excel("C:/Users/Robin Ochieng/OneDrive - Kenbright/Attachments/projects/2024/April/Discounting Automation/Data/Data.xlsx", 
                                        sheet_name = "NSE_Yield_Curve_2023")


# Define a function to check the status based on the claim number
def check_status(claim_no):
    return 'Paid/ Closed' if claim_no in claim_numbers_2022 else 'Still_Outstanding'

# Apply the function to each row of data_OS_2021
data_OS_2021['Status'] = data_OS_2021['ClaimNo'].apply(check_status)

# Calculate claim indicator count, total OS amount and percentage for 2021
claim_indicator_count_2021 = data_OS_2021.groupby('Claim Indicator').agg({'OS Amount': ['count', 'sum']})
claim_indicator_count_2021.columns = ['count', 'total_OS_Amount']
claim_indicator_count_2021['percentage'] = claim_indicator_count_2021['total_OS_Amount'] / claim_indicator_count_2021['total_OS_Amount'].sum() * 100

# Calculate claim indicator count, total OS amount and percentage for 2022
claim_indicator_count_2022 = data_OS_2022.groupby('Claim Indicator').agg({'OS Amount': ['count', 'sum']})
claim_indicator_count_2022.columns = ['count', 'total_OS_Amount']
claim_indicator_count_2022['percentage'] = claim_indicator_count_2022['count'] / claim_indicator_count_2022['count'].sum() * 100

# Calculate total OS amount summary by Class for 2022 where Claim Indicator is 'E1'
claim_indicator_sum_2022 = data_OS_2022[data_OS_2022['Claim Indicator'] == 'E1'].groupby('IRA Class')['OS Amount'].sum()

# Calculate count and total Gross Paid for Paid Claims in 2022
Paid_claims_indicator_count_sum_2022 = data_Paid_2022.groupby('Claim Indicator').agg({'Gross Paid': ['count', 'sum']})
Paid_claims_indicator_count_sum_2022.columns = ['count', 'total_Gross_Paid']

# Calculate count and total Gross Paid for Paid Claims in 2022 by Final Claim Indicator
Paid_Final_claims_indicator_count_sum_2022 = data_Paid_2022.groupby('Final Claim Indicator').agg({'Gross Paid': ['count', 'sum']})
Paid_Final_claims_indicator_count_sum_2022.columns = ['count', 'total_Gross_Paid']

# Calculate count and total Gross Paid for Paid Claims in 2022 by Loss Year where Revised Claim Indicator is 'B2'
Paid_claims_Loss_year_count_sum_2022 = data_Paid_2022[data_Paid_2022['Revised Claim Indicator'] == 'B2'].groupby('Loss Year')['Gross Paid'].agg(['count', 'sum'])

# Calculate total OS amount for OS claims by Loss Year for 2022
OS_claims_by_Loss_Year_2022 = data_OS_2022.groupby('Loss Year')['OS Amount'].sum()

# Calculate total OS amount for OS claims by Loss Year for 2023
OS_claims_by_Loss_Year_2023 = data_OS_2023.groupby('Loss Year')['OS Amount'].sum()

# Calculate total OS amount for OS Claims by IRA CLASS
OS_claims_by_IRA_Class = data_OS_2023.groupby('IRA Class')['OS Amount'].sum()

# Calculate total OS amount for OS Claims by IRA CLASS for 2023
OS_claims_by_IRA_Class_2023 = data_OS_2023[data_OS_2023['Loss Year'] == 2023].groupby('IRA Class')['OS Amount'].sum()